File: taurus-protect-sdk-typescript/src/models/change.ts
/**
 * Change models for Taurus-PROTECT SDK.
 *
 * Changes represent modifications to system configuration that require approval
 * before taking effect. This includes user management, role assignments, and
 * other administrative operations that follow an approval workflow.
 */

/**
 * The status of a change request.
 */
export type ChangeStatus = 'Created' | 'Approved' | 'Rejected' | 'Canceled';

/**
 * Represents a change record in the Taurus-PROTECT audit system.
 *
 * Changes track modifications made to system entities (wallets, addresses, users, etc.)
 * providing a complete audit trail of who changed what, when, and why. Each change
 * captures the before and after state of modified fields.
 */
export interface Change {
  /** Unique identifier for this change record */
  readonly id?: string;
  /** ID of the tenant (organization) where this change occurred */
  readonly tenantId?: number;
  /** Internal user ID of the user who made the change */
  readonly creatorId?: string;
  /** External user ID (from SSO/IdP) of the user who made the change */
  readonly creatorExternalId?: string;
  /** The type of action performed (e.g., "create", "update", "delete") */
  readonly action?: string;
  /** The type of entity that was changed (e.g., "wallet", "address", "user") */
  readonly entity?: string;
  /** The numeric ID of the entity that was changed */
  readonly entityId?: string;
  /** The UUID of the entity that was changed (if applicable) */
  readonly entityUUID?: string;
  /** Map of field names to their new values after the change */
  readonly changes?: Record<string, string>;
  /** Optional comment explaining the reason for the change */
  readonly comment?: string;
  /** Timestamp when the change was made */
  readonly createdAt?: Date;
}

/**
 * Result of a change list query with cursor-based pagination.
 */
export interface ListChangesResult {
  /** The list of changes in this page */
  readonly changes: Change[];
  /** Current page cursor (base64 encoded) */
  readonly currentPage?: string;
  /** Whether there are more pages available */
  readonly hasNext: boolean;
}

/**
 * Options for listing changes.
 */
export interface ListChangesOptions {
  /**
   * The entity type to filter by.
   * Can be one of: user, group, usergroup, businessrule, exchange, price, action,
   * feepayer, userapikey, securitydomain, taurusnetworkparticipant, visibilitygroup,
   * uservisibilitygroup, manualaccountbalancefreeze, wallet, whitelistedaddress,
   * autotransfereventhandler
   */
  entity?: string;
  /** The status to filter by */
  status?: ChangeStatus;
  /** Filter by creator ID */
  creatorId?: string;
  /** Sort order (ASC or DESC, default DESC) */
  sortOrder?: 'ASC' | 'DESC';
  /** Page size (default 50) */
  pageSize?: number;
  /** Current page cursor for pagination */
  currentPage?: string;
  /** Page request type (FIRST, PREVIOUS, NEXT, LAST) */
  pageRequest?: 'FIRST' | 'PREVIOUS' | 'NEXT' | 'LAST';
  /** Filter by entity IDs (valid when entity type is given) */
  entityIDs?: string[];
  /** Filter by entity UUIDs (valid when entity type is given) */
  entityUUIDs?: string[];
}

/**
 * Options for listing changes for approval.
 */
/**
 * Request to create a configuration change.
 */
export interface CreateChangeRequest {
  /** Action type (e.g., "update", "create", "delete") */
  action: string;
  /** Entity type (e.g., "businessrule", "user", "group") */
  entity: string;
  /** ID of the entity to change */
  entityId?: string;
  /** Map of field names to their new values */
  changes?: Record<string, string>;
  /** Optional description of the change request */
  comment?: string;
}

/**
 * Options for listing changes for approval.
 */
export interface ListChangesForApprovalOptions {
  /** Entity types to filter by */
  entities?: string[];
  /** Sort order (ASC or DESC, default DESC) */
  sortOrder?: 'ASC' | 'DESC';
  /** Page size (default 50) */
  pageSize?: number;
  /** Current page cursor for pagination */
  currentPage?: string;
  /** Page request type (FIRST, PREVIOUS, NEXT, LAST) */
  pageRequest?: 'FIRST' | 'PREVIOUS' | 'NEXT' | 'LAST';
  /** Filter by entity IDs (valid when one entity type is given) */
  entityIDs?: string[];
  /** Filter by entity UUIDs (valid when one entity type is given) */
  entityUUIDs?: string[];
}

File: taurus-protect-sdk-java/client/src/main/java/com/taurushq/sdk/protect/client/model/Change.java
package com.taurushq.sdk.protect.client.model;

import org.apache.commons.lang3.builder.ToStringBuilder;

import java.time.OffsetDateTime;
import java.util.Map;

/**
 * Represents a change record in the Taurus Protect audit system.
 * <p>
 * Changes track modifications made to system entities (wallets, addresses, users, etc.)
 * providing a complete audit trail of who changed what, when, and why. Each change
 * captures the before and after state of modified fields.
 *
 * @see ChangeResult
 */
public class Change {

    /**
     * Unique identifier for this change record.
     */
    private String id;

    /**
     * ID of the tenant (organization) where this change occurred.
     */
    private int tenantId;

    /**
     * Internal user ID of the user who made the change.
     */
    private String creatorId;

    /**
     * External user ID (from SSO/IdP) of the user who made the change.
     */
    private String creatorExternalId;

    /**
     * The type of action performed (e.g., "create", "update", "delete").
     */
    private String action;

    /**
     * The type of entity that was changed (e.g., "wallet", "address", "user").
     */
    private String entity;

    /**
     * The numeric ID of the entity that was changed.
     */
    private String entityId;

    /**
     * The UUID of the entity that was changed (if applicable).
     */
    private String entityUUID;

    /**
     * Map of field names to their new values after the change.
     */
    private Map<String, String> changes;

    /**
     * Optional comment explaining the reason for the change.
     */
    private String comment;

    /**
     * Timestamp when the change was made.
     */
    private OffsetDateTime createdAt;

    @Override
    public String toString() {
        return ToStringBuilder.reflectionToString(this);
    }

    /**
     * Returns the unique identifier for this change record.
     *
     * @return the change record ID
     */
    public String getId() {
        return id;
    }

    /**
     * Sets the unique identifier for this change record.
     *
     * @param id the change record ID to set
     */
    public void setId(String id) {
        this.id = id;
    }

    /**
     * Returns the tenant ID where this change occurred.
     *
     * @return the tenant ID
     */
    public int getTenantId() {
        return tenantId;
    }

    /**
     * Sets the tenant ID where this change occurred.
     *
     * @param tenantId the tenant ID to set
     */
    public void setTenantId(int tenantId) {
        this.tenantId = tenantId;
    }

    /**
     * Returns the internal user ID of the user who made the change.
     *
     * @return the creator's user ID
     */
    public String getCreatorId() {
        return creatorId;
    }

    /**
     * Sets the internal user ID of the user who made the change.
     *
     * @param creatorId the creator's user ID to set
     */
    public void setCreatorId(String creatorId) {
        this.creatorId = creatorId;
    }

    /**
     * Returns the external user ID (from SSO/IdP) of the user who made the change.
     *
     * @return the creator's external user ID, or {@code null} if not set
     */
    public String getCreatorExternalId() {
        return creatorExternalId;
    }

    /**
     * Sets the external user ID (from SSO/IdP) of the user who made the change.
     *
     * @param creatorExternalId the creator's external user ID to set
     */
    public void setCreatorExternalId(String creatorExternalId) {
        this.creatorExternalId = creatorExternalId;
    }

    /**
     * Returns the type of action performed.
     *
     * @return the action (e.g., "create", "update", "delete")
     */
    public String getAction() {
        return action;
    }

    /**
     * Sets the type of action performed.
     *
     * @param action the action to set
     */
    public void setAction(String action) {
        this.action = action;
    }

    /**
     * Returns the type of entity that was changed.
     *
     * @return the entity type (e.g., "wallet", "address", "user")
     */
    public String getEntity() {
        return entity;
    }

    /**
     * Sets the type of entity that was changed.
     *
     * @param entity the entity type to set
     */
    public void setEntity(String entity) {
        this.entity = entity;
    }

    /**
     * Returns the numeric ID of the entity that was changed.
     *
     * @return the entity ID
     */
    public String getEntityId() {
        return entityId;
    }

    /**
     * Sets the numeric ID of the entity that was changed.
     *
     * @param entityId the entity ID to set
     */
    public void setEntityId(String entityId) {
        this.entityId = entityId;
    }

    /**
     * Returns the UUID of the entity that was changed.
     *
     * @return the entity UUID, or {@code null} if not applicable
     */
    public String getEntityUUID() {
        return entityUUID;
    }

    /**
     * Sets the UUID of the entity that was changed.
     *
     * @param entityUUID the entity UUID to set
     */
    public void setEntityUUID(String entityUUID) {
        this.entityUUID = entityUUID;
    }

    /**
     * Returns the map of field names to their new values after the change.
     *
     * @return the changes map
     */
    public Map<String, String> getChanges() {
        return changes;
    }

    /**
     * Sets the map of field names to their new values after the change.
     *
     * @param changes the changes map to set
     */
    public void setChanges(Map<String, String> changes) {
        this.changes = changes;
    }

    /**
     * Returns the optional comment explaining the reason for the change.
     *
     * @return the comment, or {@code null} if none provided
     */
    public String getComment() {
        return comment;
    }

    /**
     * Sets the optional comment explaining the reason for the change.
     *
     * @param comment the comment to set
     */
    public void setComment(String comment) {
        this.comment = comment;
    }

    /**
     * Returns the timestamp when the change was made.
     *
     * @return the creation timestamp
     */
    public OffsetDateTime getCreatedAt() {
        return createdAt;
    }

    /**
     * Sets the timestamp when the change was made.
     *
     * @param createdAt the creation timestamp to set
     */
    public void setCreatedAt(OffsetDateTime createdAt) {
        this.createdAt = createdAt;
    }
}
